google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java

```diff
@@ -72,45 +75,60 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
       this.lastUpdateTimeInSecond = new AtomicLong(0);
     }
 
     synchronized void update(long value, long timestampMs) {
+      updateStartTime(timestampMs);
+
       long now = TimeUnit.MILLISECONDS.toSeconds(timestampMs);
-      Preconditions.checkArgument(
-          now >= lastUpdateTimeInSecond.get(), "can't update an event in the past");
-      if (lastUpdateTimeInSecond.get() == 0) {
+      long elapsed = now - startTimeSecond;
+      double weight = getWeight(elapsed);
+      // Exponential moving average = weightedSum / weightedCount, where
+      // weightedSum(n) = weight(n) * value(n) + weightedSum(n - 1)
+      // weightedCount(n) = weight(n) + weightedCount(n - 1), where weight(n) grows exponentially
+      // over elapsed time.
+      // Using weighted count in case the sum overflows.
+      mean =
+          mean * (weightedCount / (weightedCount + weight))
+              + weight * value / (weightedCount + weight);
+      weightedCount += weight;
+
+      // Set last update time so when we're getting the mean we can calculate the decay based on the
+      // last time the mean was updated.
+      if (now > lastUpdateTimeInSecond.get()) {
         lastUpdateTimeInSecond.set(now);
-        mean = value;
-        weightedCount = 1;
-      } else {
-        long prev = lastUpdateTimeInSecond.getAndSet(now);
-        long elapsed = now - prev;
-        double alpha = getAlpha(elapsed);
-        // Exponential moving average = weightedSum / weightedCount, where
-        // weightedSum(n) = value + alpha * weightedSum(n - 1)
-        // weightedCount(n) = 1 + alpha * weightedCount(n - 1)
-        // Using weighted count in case the sum overflows
-        mean =
-            mean * ((weightedCount * alpha) / (weightedCount * alpha + 1))
-                + value / (weightedCount * alpha + 1);
-        weightedCount = weightedCount * alpha + 1;
       }
     }
 
     double getMean(long timestampMs) {
-      long timestampSecs = TimeUnit.MILLISECONDS.toSeconds(timestampMs);
-      long elapsed = timestampSecs - lastUpdateTimeInSecond.get();
-      return mean * getAlpha(Math.max(0, elapsed));
+      long timestampSecond = TimeUnit.MILLISECONDS.toSeconds(timestampMs);
+      long elapsed = timestampSecond - lastUpdateTimeInSecond.get();
+      double decay = getWeight(Math.min(0, -elapsed));
+      return mean * decay;
+    }
+
+    private double getWeight(long elapsedSecond) {
+      return Math.exp(decayConstant * elapsedSecond);
     }
 
-    private double getAlpha(long elapsedSecond) {
-      return Math.exp(-decayConstant * elapsedSecond);
+    private synchronized void updateStartTime(long timestampMs) {
+      long timestampSecond = TimeUnit.MILLISECONDS.toSeconds(timestampMs);
+      long elapsed = timestampSecond - lastUpdateTimeInSecond.get();
+      if (elapsed > UPDATE_START_TIME_THRESHOLD_SECOND) {
+        double decay = getWeight(-elapsed);
+        mean *= decay;
```

Comment:
nit. the negative exponent is easy to miss, it might be easier to drop the negative exponent and change this to a division.